### ISSUE TITLE
Resolve dependencies recursively

### ISSUE THREAD

**Issue by alanchchen** [kubernetes/helm#1969]
`Helm 2.2.0`

I have charts as following
```
all-my-charts
my-chart-1
my-chart-2
my-chart-3
my-base-chart
```

`my-chart-1`, `my-chart-2` and `my-chart-3` require `my-base-chart`
`all-my-charts` is an empty chart but it requires `my-chart-1`, `my-chart-2` and `my-chart-3`

When I run
```
$ helm install -n my-release all-my-charts
```

There're warnings if I didn't run `helm dep up all-my-charts`

> Warning: my-chart-1 is in requirements.yaml but not in the charts/ directory!

However the `helm dep up all-my-charts` didn't update dependencies for `my-chart-1`, `my-chart-2` and `my-chart-3` and there is no warning.

**Comment by technosophos** [comment]
I would be okay with implementing a very limited version of this feature with a `--recursive` flag on `helm dep up`. But this was not the intended usage of this command.

To implement proper recursive dependency resolution, yet remain compliant and backward compatible with Tiller, we'd need to:

- Hide the behavior behind a `--recursive` flag (to not break existing scripts)
- **NOT** run dependency resolution on subcharts that are already packaged as tgz files
- On subcharts that are directories, run the `helm dep up` logic in that directory

Packaged charts should never have their dependencies updated, as they are considered immutable once they are packaged. Updating them will (1) harm the reproducibility of a build, and (2) break the package signing system.

Again, this is straying a little bit from the original intentions for `requirements.yaml`, which was to ONLY import tgz files that had already been packaged. This matches an OS package manager's behavior (require pre-built existing packages, don't rebuild packages on the fly), which was the design intent of Helm.

(What I would really suggest is simply writing a Helm plugin that does the recursive build.)

**Comment by alanchchen** [comment]
I think this feature might need more discussion.
To make things simpler, writing a plugin would be better.
Thanks @technosophos .

**Comment by technosophos** [comment]
I thought about this a lot last night. With the addition of the `file://` protocol that @qwangrepos did for 2.2.0, it actually might make sense to do some recursive building on any chart that uses the `file://` protocol.

@qwangrepos if you have a moment to offer an opinion on this, it would be helpful.

Basically, I don't want to break backward compatibility, but if we could safely improve the developer experience as @alanchchen suggests, we should do it.

**Comment by qwangrepos** [comment]
Current support of `file://` protocol compresses the subchart directory into a tgz files and save it into the `charts/` folder. This can be twisted to save the subchart directory. If the subchart has a requirements.yaml, run the `helm dep up` logic for the subchart. It continues until all dependencies have been resolved. 
For it is supported by using `file://` for chart development, it shouldn't touch other parts of the dependency update logic from what I am seeing. It should be okay for backward compatibility. 
This will change `helm dep up` to recursively build subcharts into `charts/` folder of current chart and subcharts. 
This is a use case that hasn't been considered in the `file://` support yet but I think it is necessary to add it on.

**Comment by wilkers-steve** [comment]
I think this is better suited as a plugin.  It feels very shortsighted to me to only apply this logic to one particular method of referring to requirements (file:// versus declaring required charts in a repository).  A plugin address this particular use case perfectly, and implementing a plugin doesn't add additional conditional debt to Helm by applying to one specific method of declaring dependencies.

**Comment by thomastaylor312** [comment]
This has a proposal that was accepted and is now a feature request being implemented in #2247. So I am closing this issue
